src/test/java/org/jenkinsci/plugins/workflow/GraphListenerTest.java

```diff
@@ -0,0 +1,50 @@
+package org.jenkinsci.plugins.workflow;
+
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.flow.GraphListener;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+
+import java.io.Serializable;
+import java.util.Random;
+
+public class GraphListenerTest
+{
+    @ClassRule
+    public static JenkinsRule r = new JenkinsRule();
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Test
+    public void listener() throws Exception {
+        String script = "node { \n" //
+            + "    echo \"hello\"\n" //
+            + "  "
+            + "}";
+        WorkflowJob j = r.createProject(WorkflowJob.class, "listener");
+        j.setDefinition(new CpsFlowDefinition(script, true));
+        r.buildAndAssertSuccess(j);
+    }
+
+    @TestExtension
+    public static class TestGraphListener implements GraphListener, Serializable
+    {
+        private Random random = new Random();
```

Comment:
nit: Seems better to avoid randomness in the test, maybe by throwing an exception in the listener unconditionally and asserting that the exception's message is printed in the logs.

Reply:
@dwnusbaum  good idea